openhtf/core/test_state.py

```diff
@@ -130,7 +146,52 @@ def test_api(self):
                 running_phase_state.attachments,
                 running_phase_state.attach,
                 running_phase_state.attach_from_file,
-                self.notify_update))
+                self.get_measurement,
+                self.get_attachment,
+                self.notify_update,))
+
+  def get_measurement(self, measurement_name):
+    """Get a copy of a measurement value from current or previous phase.
+
+    Args:
+      measurement_name: str of the measurement name
+    Returns:
+      an ImmutableMeasurement
+    """
+    # The framework ignores measurements from SKIP and REPEAT phases
+    ignore_outcomes = (test_record.PhaseOutcome.SKIP,)
+    # Iterate through phases in reversed order for LIFO.
+    for phase_record in reversed(self.test_record.phases):
+      if (measurement_name in phase_record.measurements and
+          phase_record.result not in ignore_outcomes):
+        measurement = phase_record.measurements[measurement_name]
+        return ImmutableMeasurement.FromMeasurement(measurement)
+
+    # Check current running phase state
+    if self.running_phase_state:
+      if measurement_name in self.running_phase_state.measurements:
+        return ImmutableMeasurement.FromMeasurement(
+            self.running_phase_state.measurements[measurement_name])
+
+    test.logger.warning('Could not find measurement: %s', measurement_name)
+    return None
+
+  def get_attachment(self, attachment_name):
+    """Get a copy of an attachment contents from current or previous phase."""
+    for phase_record in self.test_record.phases:
+      if attachment_name in phase_record.attachments:
+        attachment = phase_record.attachments[attachment_name]
+        return copy.deepcopy(attachment.data)
+
+    # Check current running phase state
+    if self.running_phase_state:
+      if attachment_name in self.running_phase_state.phase_record.attachments:
+        attachment = self.running_phase_state.phase_record.attachments.get(
+            attachment_name)
+        return copy.deepcopy(attachment.data)
```

Comment:
The data is already a (immutable in Python) string, so copying it is a waste of resources.

Also, what about the mimetype?

Reply:
Done.  Modified to have get_attachment return a copy of an Attachment, instead of just the contents.

Reply:
Agreed.